boards: toradex: add toradex verdin imx8mm board (#61047)
zephyrproject-rtos/zephyr: boards/toradex/verdin_imx8mm/doc/index.rst

> .. _verdin_imx8mm_m4:
> 
> Toradex Verdin iMX8M Mini M4
> #############################

Reviewer: ```suggestion
.. zephyr:board:: verdin_imx8mm
```

Overview
********

> .. image:: img/verdin_imx8mm_m4.webp
>    :align: center
>    :alt: VERDIN IMX8MM M4
> 

Reviewer: Image will be automatically inserted

```suggestion
```

commercial graded products and 1.6 GHz for industrial temperature range products.

In addition to the main CPU complex, the i.MX 8M Mini features a Cortex-M4F processor which
peaks up to 400 MHz. This processor is independent of the main complex and features its own
dedicated interfaces while still being able to access the regular interfaces. This heterogeneous
multicore system allows for the running of additional real-time operating systems on the M4 cores
for time- and security-critical tasks.

- Board features:

  - RAM: 1GB - 2GB (LPDDR4)
  - Storage:

    - 4GB - 17GB eMMC
    - 2KB I²C EEPROM
  - Wireless:

    - WiFi Dual-band 802.11 ac/a/b/g/n 2.4/5 GHz
    - Bluetooth 5/BLE
  - USB:

    - 1x USB2.0 Host
    - 1x USB2.0 OTG
  - Ethernet Gigabit
  - Interfaces:

    - MIPI DSI 1x4 Data Lanes
    - PCIe Gen 2
    - MIPI CSI-2
    - SPI
    - QSPI
    - UART
    - I²C
    - I²S
    - SD/SDIO/MMC
    - GPIOs
    - CAN
    - ADC
    - S/PDIF
  - Debug

    - JTAG 10-pin connector

More information about the board can be found at the
`Toradex website`_.

Supported Features
==================

> The Zephyr verdin_imx8mm_m4 board configuration supports the following hardware

Reviewer 2: use proper board name as supplied to west build with \`\` around it and call it board target

> The Zephyr :zephyr_file:`verdin_imx8mm_mimx8mm6_m4_defconfig` board target supports the following hardware

Reviewer 2: ```suggestion
The Zephyr ``verdin_imx8mm/mimx8mm6/m4`` board target supports the following hardware
```

features:

+-----------+------------+-------------------------------------+
| Interface | Controller | Driver/Component                    |
+===========+============+=====================================+
| NVIC      | on-chip    | nested vector interrupt controller  |
+-----------+------------+-------------------------------------+
| SYSTICK   | on-chip    | systick                             |
+-----------+------------+-------------------------------------+
| CLOCK     | on-chip    | clock_control                       |
+-----------+------------+-------------------------------------+
| PINMUX    | on-chip    | pinmux                              |
+-----------+------------+-------------------------------------+
| UART      | on-chip    | serial port-polling;                |
|           |            | serial port-interrupt               |
+-----------+------------+-------------------------------------+
| GPIO      | on-chip    | GPIO output                         |
|           |            | GPIO input                          |
+-----------+------------+-------------------------------------+

The default configuration can be found in the defconfig file:

> :zephyr_file:`boards/arm/verdin_imx8mm_m4/verdin_imx8mm_m4_defconfig`.

Reviewer 3: File path here needs to be updated

Reviewer 4: This has been addressed in the latest push.

It is recommended to disable peripherals used by the M4 core on the Linux host.

Other hardware features are not currently supported by the port.

Connections and IOs
===================

UART:

Zephyr is configured to use the UART4 by default, which is connected to the FTDI
USB converter on most Toradex carrier boards.

This is also the UART connected to WiFi/BT chip in modules that have the WiFi/BT
chip. Therefore, if UART4 is used, WiFI/BT will not work properly.

If the WiFi/BT is needed, then another UART should be used for Zephyr (UART2 for
example). You can change the UART by changing the ``zephyr,console`` and

> ``zephyr,shell-uart`` in the :zephyr_file:`boards/arm/verdin_imx8mm_m4.dts` file.

Reviewer 3: File path needs updating

Reviewer 4: This has been addressed in the latest push.

+---------------+-----------------+---------------------------+
| Board Name    | SoC Name        | Usage                     |
+===============+=================+===========================+
| UART_1        | UART2           | General purpose UART      |
+---------------+-----------------+---------------------------+
| UART_2        | UART3           | General purpose UART      |
+---------------+-----------------+---------------------------+
| UART_3        | UART1           | Cortex-A53 debug UART     |
+---------------+-----------------+---------------------------+
| UART_4        | UART4           | Cortex-M4 debug UART      |
+---------------+-----------------+---------------------------+

GPIO:

All the GPIO banks are enabled in the :zephyr_file:`dts/arm/nxp/nxp_imx8m_m4.dtsi`.

LED:

There is no LED in the module itself, this is dependent on the carrier board that
is being used with the module. The device tree is configured to use the GPIO3_IO1,
which can be connected to the LED of the Verdin Development Board or changed in the

> :zephyr_file:`boards/arm/verdin_imx8mm_m4.dts` if needed.

Reviewer 3: File path out of date

Reviewer 4: This has been addressed in the latest push.

| OCRAM      | 0x00900000-0x0093FFFF   | 0x20200000-0x2023FFFF  | 0x00900000-0x0093FFFF | 256KB                |
+------------+-------------------------+------------------------+-----------------------+----------------------+
| TCMU       | 0x00800000-0x0081FFFF   | 0x20000000-0x2001FFFF  |                       | 128KB                |
+------------+-------------------------+------------------------+-----------------------+----------------------+
| TCML       | 0x007E0000-0x007FFFFF   |                        | 0x1FFE0000-0x1FFFFFFF | 128KB                |
+------------+-------------------------+------------------------+-----------------------+----------------------+
| OCRAM_S    | 0x00180000-0x00187FFF   | 0x20180000-0x20187FFF  | 0x00180000-0x00187FFF | 32KB                 |
+------------+-------------------------+------------------------+-----------------------+----------------------+

For more information about memory mapping see the
`i.MX 8M Applications Processor Reference Manual`_  (section 2.1.2 and 2.1.3)

At compilation time you have to choose which RAM will be used. This

> configuration is done in the file ``boards/arm/verdin_imx8mm_m4/verdin_imx8mm_m4.dts``

Reviewer 3: File path out of date, also please use `zephyr_file` here

Reviewer 4: This has been addressed in the latest push.

Copy the ``zepyhr.elf`` to ``/lib/firmware`` on the target.

.. note::
   In order to use remoteproc you have to add ``imx8mm-verdin_hmp_overlay.dtbo`` at
   the end of the line in the ``/boot/overlays.txt``, then reboot the target. If
   you are using `TorizonCore`_, then this file is located at
   ``/boot/ostree/torizon-<hash>/dtb/overlays.txt``.

To load and start a firmware use these commands:

.. code-block:: console

   verdin-imx8mm:~# echo zepyhr.elf > /sys/class/remoteproc/remoteproc0/firmware
   verdin-imx8mm:~# echo start > /sys/class/remoteproc/remoteproc0/state
   [   94.714498] remoteproc remoteproc0: powering up imx-rproc
   [   94.720481] remoteproc remoteproc0: Booting fw image zephyr.elf, size 473172
   [   94.727713] remoteproc remoteproc0: No resource table in elf
   [   94.733615] remoteproc remoteproc0: remote processor imx-rproc is now up

The M4-Core is now started up and running. You can see the output from Zephyr
on UART4.

Debugging
=========

MIMX8MM EVK board can be debugged by connecting an external JLink
JTAG debugger to the J902 debug connector and to the PC. Then
the application can be debugged using the usual way.

> Here is an example for the :ref:`hello_world` application.

Reviewer: ```suggestion
Here is an example for the :zephyr:code-sample:`hello_world` application.
```

This fixes the doc build failure

.. zephyr-app-commands::
   :zephyr-app: samples/hello_world

>    :board: verdin_imx8mm_m4

Reviewer 3: Board target should be `verdin_imx8mm/mimx8mm6/m4`

Reviewer 4: This has been addressed in the latest push.
